cpp/src/community/detail/common_methods.cuh

```diff
@@ -0,0 +1,370 @@
+
+namespace cugraph {
+namespace detail {
+
+// a workaround for cudaErrorInvalidDeviceFunction error when device lambda is used
+template <typename vertex_t, typename weight_t>
+struct key_aggregated_edge_op_t {
+  weight_t total_edge_weight{};
+  weight_t resolution{};
+  __device__ auto operator()(
+    vertex_t src,
+    vertex_t neighbor_cluster,
+    weight_t new_cluster_sum,
+    thrust::tuple<weight_t, vertex_t, weight_t, weight_t, weight_t> src_info,
+    weight_t a_new) const
+  {
+    auto k_k              = thrust::get<0>(src_info);
+    auto src_cluster      = thrust::get<1>(src_info);
+    auto a_old            = thrust::get<2>(src_info);
+    auto old_cluster_sum  = thrust::get<3>(src_info);
+    auto cluster_subtract = thrust::get<4>(src_info);
+
+    if (src_cluster == neighbor_cluster) new_cluster_sum -= cluster_subtract;
+
+    weight_t delta_modularity = 2 * (((new_cluster_sum - old_cluster_sum) / total_edge_weight) -
+                                     resolution * (a_new * k_k - a_old * k_k + k_k * k_k) /
+                                       (total_edge_weight * total_edge_weight));
+
+    return thrust::make_tuple(neighbor_cluster, delta_modularity);
+  }
+};
+
+// a workaround for cudaErrorInvalidDeviceFunction error when device lambda is used
+template <typename vertex_t, typename weight_t>
+struct reduce_op_t {
+  using type                          = thrust::tuple<vertex_t, weight_t>;
+  static constexpr bool pure_function = true;  // this can be called from any process
+  inline static type const identity_element =
+    thrust::make_tuple(std::numeric_limits<weight_t>::lowest(), invalid_vertex_id<vertex_t>::value);
+
+  __device__ auto operator()(thrust::tuple<vertex_t, weight_t> p0,
+                             thrust::tuple<vertex_t, weight_t> p1) const
+  {
+    auto id0 = thrust::get<0>(p0);
+    auto id1 = thrust::get<0>(p1);
+    auto wt0 = thrust::get<1>(p0);
+    auto wt1 = thrust::get<1>(p1);
+
+    return (wt0 < wt1) ? p1 : ((wt0 > wt1) ? p0 : ((id0 < id1) ? p0 : p1));
+  }
+};
+
+// a workaround for cudaErrorInvalidDeviceFunction error when device lambda is used
+template <typename vertex_t, typename weight_t>
+struct cluster_update_op_t {
+  bool up_down{};
+  __device__ auto operator()(vertex_t old_cluster, thrust::tuple<vertex_t, weight_t> p) const
+  {
+    vertex_t new_cluster      = thrust::get<0>(p);
+    weight_t delta_modularity = thrust::get<1>(p);
+
+    return (delta_modularity > weight_t{0})
+             ? (((new_cluster > old_cluster) != up_down) ? old_cluster : new_cluster)
+             : old_cluster;
+  }
+};
+
+// a workaround for cudaErrorInvalidDeviceFunction error when device lambda is used
+template <typename vertex_t, typename weight_t>
+struct return_edge_weight_t {
+  __device__ auto operator()(
+    vertex_t, vertex_t, weight_t w, thrust::nullopt_t, thrust::nullopt_t) const
+  {
+    return w;
+  }
+};
+
+template <typename graph_view_t>
```

Comment:
Verbosity more than anything.

I tend to prefer `template <typename vertex_t, typename edge_t, typename weight_t, bool multi_gpu>`.  But the edge_src_property_t was the graph_view_type.  So that would require a couple of extra lines of parameter definition, using `graph_view_t` is a bit clearer in that case.

I can switch to `graph_view_t` for detail methods for consistency if that's what we think is better.